### PR TITLE
Add early stopping and modelcheckpoint callbacks

### LINKED ISSUE #150: Add validation set and callbacks in pretrain and train scripts
Currently, we only have train accuracy and its not being validated. We can go for a simple validation script or a K fold validation. 
With validation check , we can add callbacks such as early stopping and model checkpoints.

### PR DESCRIPTION
**Pull request by osmarcedron**
Fixes #150.

### CODE REVIEW

File: pretrain.py [hunk 1]
```diff
@@ -74,11 +76,17 @@ def gen_func(images):
 model = network.create(input_shape, args.network)
 network.compile(model, args.optimizer)
 
-model.save('pretrain.h5')
+# serialize model to JSON
```
Comment (marco-c) [inline pretrain.py]: Why did you change this?

File: pretrain.py [hunk 2]
```diff
@@ -74,11 +76,17 @@ def gen_func(images):
...
+model_json = model.to_json()
+with open('model_pretrain.json', 'w') as json_file:
+    json_file.write(model_json)
 
-model.fit_generator(train_iterator, validation_data=validation_iterator, steps_per_epoch=train_couples_len / BATCH_SIZE, validation_steps=validation_couples_len / BATCH_SIZE, epochs=EPOCHS)
+filepath = 'best_pretrain_weights.hdf5'
```
Comment (marco-c) [inline pretrain.py]: Remove this variable and use the value directly in the `ModelCheckpoint` constructor.

File: pretrain.py [hunk 3]
```diff
@@ -74,11 +76,17 @@ def gen_func(images):
...
+early_stopping = EarlyStopping(monitor='val_accuracy', patience=2)
```
Comment (marco-c) [inline pretrain.py]: Early stopping should be an option of the scripts.

File: train.py [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import argparse
```
Comment (marco-c) [inline train.py]: Same comments apply to this file.

File: pretrain.py (hunk 1 above)
Comment (osmarcedron) [reply pretrain.py]: As the ModelCheckpoint callback saves the full model, something similar to what model.save does, I thought the latter would be redundant.

Comment (marco-c) [reply pretrain.py]: That sounds good, but why do we need to save the model as JSON? We probably can just remove the saving at the beginning altogether, no?

Comment (osmarcedron) [reply pretrain.py]: @marco-c. Indeed, that was a mistake and it is already been removed, please check it now.

File: README.md [hunk 5]
```diff
@@ -79,6 +79,8 @@ For the unsupervised training, we are using a related problem for which we alrea
     -optimizer                To select the optimizer to use   
                               
     -classification_type      Either Y vs N + D or Y + N vs D
+
+    --early_stoppping	      (Optional) To stop training when val_accuracy has stopped improving
```
Comment (marco-c) [inline README.md]: Let's use the full wording (`validation accuracy`) as it's clearer.

File: pretrain.py [hunk 6]
```diff
@@ -14,6 +16,7 @@
 parser = argparse.ArgumentParser()
 parser.add_argument('network', type=str, choices=network.SUPPORTED_NETWORKS, help='Select the network to use for training')
 parser.add_argument('optimizer', type=str, choices=network.SUPPORTED_OPTIMIZERS, help='Select the optimizer to use for training')
+parser.add_argument('--early_stopping', dest='early_stopping', action='store_true', help='Stop training when a monitored quantity has stopped improving.')
```
Comment (marco-c) [inline pretrain.py]: Here let's use the same text as the description in the README

Comment (marco-c) [reply pretrain.py]: Without `To`

**Comment by osmarcedron** [comment]
@marco-c, the comments were addressed. I think Travis is having problems trying to clone the submodules.

**Comment by marco-c** [comment]
> I think Travis is having problems trying to clone the submodules.

Yep, I'll respin the build tomorrow. Hopefully the problem is transient.
